### ISSUE TITLE
[api-extractor] Does not work for export {default as x} from "x" .

### ISSUE THREAD

**Issue by Spanoudakhs** [microsoft/rushstack#1904]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

The d.ts bundling does not work if the `index.d.ts` contains `export {default as x} from "x" ;`

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

Steps to reproduce the bug :

create a project directory and in it execute :

```bash
npm init -y;
npm install --save-dev typescript @microsoft/api-extractor;
```
create `./src/fn1.ts` :
```ts
export default function() {}
```
create `./src/fn2.ts` :
```ts
export default function foo() {}
```
create `./src/index.ts` :
```ts
import fn1 from "./fn1";
import fn2 from "./fn2";
export {default as fn1} from "./fn1";
export {default as fn2} from "./fn2";
export default {fn1,fn2};
```
create `./tsconfig,json` :
```json
{
	"compilerOptions": {
		"module": "ESNext",
		"declaration": true,
		"target": "ESNext",
		"rootDir": "./src",
		"outDir": "./built",
	},
	"include": [
		"./src"
	]
}
```
creates `./api-extractor.json` :
```json
{
  "mainEntryPointFilePath": "./built/index.d.ts",
  "bundledPackages": [ ],
  "compiler": {
    "tsconfigFilePath": "./tsconfig.json"
  },
  "apiReport": {
    "enabled": false
  },
  "docModel": {
    "enabled": false
  },
  "dtsRollup": {
    "enabled": true,
    "untrimmedFilePath": "./dist/typings/index.d.ts"
  },
  "tsdocMetadata": {
    "enabled": false
  },
  "newlineKind": "os"
}
```
the project folder now should look like this :
```text
project-folder
├─node_modules
├─src
│  ├─index.ts
│  ├─fn1.ts
│  └─fn2.ts
├─api-extractor.json
├─package.json
└─tsconfig.json
```
then excute :
```bash
npx tsc;
npx api-extractor run --local --verbose;
```
then take a look at the file `./dist/typings/index.d.ts` .

**What is the expected behavior?**

`./dist/typings/index.d.ts` should not contain mistakes

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** api-extractor
* **Tool Version:** 7.8.9
* **Node Version:** v12.6.0
  * **Is this a LTS version?** I think no
  * **Have you tested on a LTS version?** I think no
* **OS:**  ubuntu 18.04 lts bionic beaver

**Comment by octogonz** [comment]
These are deprecated practices:

```ts
export default function foo() {}
```

`export default` is not a good way to export this function, because the name cannot be checked by the compiler.  For example, if a person does `import fn2 from "./fn1";` by accident, the compiler has no way to detect the mismatched names, and the resulting error messages may be very confusing.  

Unless you're supporting a legacy library contract, you can always use `export` instead of `export default` for modern code.

```ts
export default function() {}
```

`export default function` is even worse -- how do we document something that has no name? :-)

**Comment by octogonz** [comment]
I would also consider this a deprecated pattern. 🙂
```ts
export {default as fn1} from "./fn1";
export {default as fn2} from "./fn2";
export default {fn1,fn2};
```

The compiler is forced to introduce a temporary symbol `_default`:
```ts
export { default as fn1 } from "./fn1";
export { default as fn2 } from "./fn2";
declare const _default: {
    fn1: () => void;
    fn2: () => void;
};
export default _default;
```

It's unclear how to write TSDoc comments for such a construct.  

For reference, this is  the more conventional way to represent this:

`./src/index.ts`
```ts
export { fn1 } from "./fn1";
export { fn2 } from "./fn2";
```

I'm curious, what is the reason for all these `default` exports? Is it to maintain compatibility with an older legacy release?

**Comment by octogonz** [comment]
When I tried this repro, API Extractor's .d.ts rollup output looks like this:
```
declare const _default: {
    fn1: () => void;
    fn2: () => void;
};
export default _default;

export declare function (): void;  // <====== incorrect

export declare function fn2(): void;

export { }
```

@Spanoudakhs  The `fn1` API is missing a name, but otherwise everything is correct as far I can tell.  You mentioned "mistakes" -- is there anything else incorrect?

**Comment by octogonz** [comment]
## Workaround

Here is a workaround that does not alter the exported API contract:

create `./src/fn1.ts` :
```ts
export function fn1() {}
```
create `./src/fn2.ts` :
```ts
export function foo() {}
```
create `./src/index.ts` :
```ts
import { fn1 } from "./fn1";
import { foo as fn2 } from "./fn2";
export { fn1, fn2 }
export default {fn1, fn2};
```

**Comment by octogonz** [comment]
Since there is an easy workaround, and since the usage of `default` here is a deprecated and uncommon practice, I won't have time to fix this bug.

However we will accept a PR if someone from the community wants to contribute a fix.  

For reference, here is a previous PR that made fixes for similar issues with `default` exports:  https://github.com/microsoft/rushstack/pull/1267

**Comment by Spanoudakhs** [comment]
>I'm curious, what is the reason for all these default exports? 

There are two reasons that made me open this issue :

1. I am writing three personal libraries which I will publish in npm in the future . One for es (i.e. pure es , no browser no node), one for browser and one for node . Here is how the project folder of each library looks :
```
    project-folder
    ├─src
    │ └─modules
    │   ├─fn1.ts
    │   └─fn2.ts
    ├─package.json
    └─tsconfig.json
```
Each `fn*.ts` **default exports** a single function which is intended for public usage.

I have created a node js script which I execute and it creates the file `./src/index.ts` which looks something like this :
```ts
import fn1 from "./modules/fn1";
export {default as fn1} from "./modules/fn1";

import fn2 from "./modules/fn2";
export {default as fn2} from "./modules/fn2";

export default {fn1,fn2};
```
The node js script decides about the names of the imported and exported functions by their filename . Like that it requires less work from when doing it with named exports , where I need always to check that the filename matches the name of the exported function .

But that is not a big issue since I can change the default exports to named exports .

2. I am looking for a way to make a module that does this :

```
node_modules        web_modules
┣ lit-html          ┣ lit-html.js  
┣ mobx              ┣ lit-html.d.ts
┗ mobx-utils   =>   ┣ mobx.js
                    ┣ mobx.d.ts
                    ┣ mobx-utils.js
                    ┗ mobx-utils.d.ts
```
i.e. converts node modules to es modules with intellisense . Like this I can avoid bundling during development stage of a web app . Nobody guarantees that in a random npm package which I will need for my web app , default exports will not be used .

Also regarding :

>the usage of default here is a deprecated

can you please point me to an appropriate source that shows that default exports are deprecated ? Do you mean that they are an anti pattern?

>You mentioned "mistakes" -- is there anything else incorrect?

No just the missing name `fn1`. 

By the way why there is an empty object exported ?

**Comment by octogonz** [comment]
> > the usage of default here is a deprecated
> 
> can you please point me to an appropriate source that shows that default exports are deprecated ? Do you mean that they are an anti pattern?

Yes, sorry that my phrasing was not more clear. I did not mean that they are officially deprecated by a standards body. Merely that people generally do not use this feature any more, because there is now a better way.

> No , just the missing name `fn1`.
> 
> By the way why there is an empty object exported ?

For a given declaration, API Extractor tracks a [local name](https://github.com/microsoft/rushstack/blob/a57573d44214d5283a166e36036f7920f4c21a36/apps/api-extractor/src/analyzer/AstSymbol.ts#L64) (i.e. the original declaration) and one or more [export names](https://github.com/microsoft/rushstack/blob/a57573d44214d5283a166e36036f7920f4c21a36/apps/api-extractor/src/collector/CollectorEntity.ts#L63) (i.e. that consumers of your library will import).  Seems like the local name is not being assigned because the function is anonymous.
